PHPCompatibility/Sniffs/Classes/NewClassesSniff.php

```diff
@@ -1373,6 +1390,38 @@ private function processCatchToken(File $phpcsFile, $stackPtr)
         }
     }
 
+    /**
+     * Processes this test for when a use token is encountered.
+     *
+     * - Save imported classes for later use.
+     *
+     * @since 10.0.0
+     *
+     * @param \PHP_CodeSniffer\Files\File $phpcsFile The file being scanned.
+     * @param int                         $stackPtr  The position of the current token in
+     *                                               the stack passed in $tokens.
+     *
+     * @return void
+     */
+    private function processUseToken(File $phpcsFile, $stackPtr)
+    {
+        if (!UseStatements::isImportUse($phpcsFile, $stackPtr)) {
+            return;
+        }
+
+        $splitUseStatement = UseStatements::splitImportUseStatement($phpcsFile, $stackPtr);
+
+        foreach ($splitUseStatement['name'] as $name => $fullyQualifiedName) {
+
+            $lowerFullyQualifiedName = strtolower($fullyQualifiedName);
+            if (isset($this->newClasses[$lowerFullyQualifiedName]) || isset($this->newExceptions[$lowerFullyQualifiedName])) {
+                continue;
```

Comment:
I tried to come up with a solution. All my solutions made the sniff slower and capped the memory usage at `count($this->newClasses) * sniffed files`.

```php
private function processUseToken(File $phpcsFile, $stackPtr)
    {
        if (!UseStatements::isImportUse($phpcsFile, $stackPtr)) {
            return;
        }

        $splitUseStatement = UseStatements::splitImportUseStatement($phpcsFile, $stackPtr);

        foreach ($splitUseStatement['name'] as $name => $fullyQualifiedName) {

            $lowerFullyQualifiedName = strtolower($fullyQualifiedName);

            foreach (array_keys($this->newClasses) as $newClass) {
                $newClassLength = strlen($newClass);
                $lowerFullyQualifiedNameLength = strlen($lowerFullyQualifiedName);
                if (
                    $newClassLength !== $lowerFullyQualifiedNameLength &&
                    substr($lowerFullyQualifiedName, -$newClassLength) === $newClass)
                {
                    $this->importedClasses[$phpcsFile->getFilename()][strtolower($name)] = true;
                    break;
                }
            }
        }
    }
```
This change runs the `PHPCompatibility/Tests/Classes/NewClassesUnitTest.php` in 00:00.717 instead of 00:00.705 on my machine.

Alternatively I could unset information about imports of other files as a new file is sniffed but I don't know enough about PHPCS to tell if this is a save operation or invites unwanted side effects.

Reply:
> Alternatively I could unset information about imports of other files as a new file is sniffed but I don't know enough about PHPCS to tell if this is a save operation or invites unwanted side effects.

In my experience, that is a safe operation, _unless_ a sniff needs to retain information cross-file (like the PHPCS native sniff checking for duplicate classes). I've looked into this before and the only potential problem I saw was with `parallel` processing, but based on the implementation details of that, I ended up concluding that it should still work fine, even when `parallel` is used (though I've never been able to properly test this as I'm on Windows).

If you need a code example for resetting on each file, have a look at the [`Universal.Namespaces.OneDeclarationPerFile` sniff](https://github.com/PHPCSStandards/PHPCSExtra/blob/develop/Universal/Sniffs/Namespaces/OneDeclarationPerFileSniff.php), where this principle is also used.
If you look at the tests for that sniff, you will also see an extra test which safeguards that the reset is executed and that there is no cross-file bleed through (like I suggested in https://github.com/PHPCompatibility/PHPCompatibility/pull/1695#discussion_r1540360928 ).

Reply:
> This change runs the PHPCompatibility/Tests/Classes/NewClassesUnitTest.php in 00:00.717 instead of 00:00.705 on my machine.

I've ran some tests with running this current patch + the code posted above over a codebase which uses import use statements extensively (PHPUnit) and I concur that not caching the statements which don't overlap, doesn't seem to make a difference.

I've also tried with another code variation and got the same result again.

In case you are wondering, this is the code sample I did the additional test run with:
```php
    private function processUseToken(File $phpcsFile, $stackPtr)
    {
        if (!UseStatements::isImportUse($phpcsFile, $stackPtr)) {
            return;
        }

        $splitUseStatement = UseStatements::splitImportUseStatement($phpcsFile, $stackPtr);
        $nameStatements    = array_change_key_case($splitUseStatement['name'], CASE_LOWER);

        foreach ($nameStatements as $name => $fullyQualifiedName) {
            $lowerFullyQualifiedName = strtolower($fullyQualifiedName);
            if (isset($this->newClasses[$lowerFullyQualifiedName])
               && isset($this->newClasses[$name]) === false
            ) {
                continue;
            }

            $this->importedClasses[$phpcsFile->getFilename()][$name] = true;
        }
    }
```

Reply:
After I check for the leak in the test and have implemented the reset, I would, with your permission, assume that the thread is resolved.

Reply:
It is indeed.